File: examples/websocket_client.py
#!/usr/bin/env python3
"""
WandererKills WebSocket Client Example (Python)

This example shows how to connect to the WandererKills WebSocket API
to receive real-time killmail updates for specific EVE Online systems.

Dependencies:
    pip install websockets asyncio json
"""

import asyncio
import json
import logging
import signal
import sys
from typing import List, Optional, Set
import websockets
from websockets.exceptions import ConnectionClosed, WebSocketException

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class WandererKillsClient:
    """WebSocket client for WandererKills real-time killmail subscriptions."""
    
    def __init__(self, server_url: str):
        self.server_url = server_url.replace('http://', 'ws://').replace('https://', 'wss://')
        self.websocket: Optional[websockets.WebSocketServerProtocol] = None
        self.subscriptions: Set[int] = set()
        self.subscription_id: Optional[str] = None
        self.running = False
        
    async def connect(self) -> dict:
        """Connect to the WebSocket server and join the killmails channel."""
        try:
            # Establish WebSocket connection with optional client identifier
            # The client_identifier helps with debugging and appears in server logs
            client_id = "python_example_client"
            uri = f"{self.server_url}/socket/websocket?vsn=2.0.0&client_identifier={client_id}"
            logger.info(f"Connecting to {uri}")
            
            self.websocket = await websockets.connect(uri)
            self.running = True
            
            # Join the killmails channel
            join_message = {
                "topic": "killmails:lobby",
                "event": "phx_join",
                "payload": {},
                "ref": 1
            }
            
            await self.websocket.send(json.dumps(join_message))
            
            # Wait for join confirmation
            response = await self.websocket.recv()
            response_data = json.loads(response)
            
            if response_data.get("event") == "phx_reply" and response_data.get("payload", {}).get("status") == "ok":
                payload = response_data["payload"]["response"]
                self.subscription_id = payload.get("subscription_id")
                logger.info(f"✅ Connected to WandererKills WebSocket")
                logger.info(f"📋 Subscription ID: {self.subscription_id}")
                
                # Start listening for messages
                asyncio.create_task(self._listen_for_messages())
                
                return payload
            else:
                raise Exception(f"Failed to join channel: {response_data}")
                
        except Exception as e:
            logger.error(f"❌ Connection failed: {e}")
            raise
    
    async def _listen_for_messages(self):
        """Listen for incoming WebSocket messages."""
        try:
            while self.running and self.websocket:
                try:
                    message = await self.websocket.recv()
                    await self._handle_message(json.loads(message))
                except ConnectionClosed:
                    logger.warning("📡 WebSocket connection closed")
                    break
                except json.JSONDecodeError as e:
                    logger.error(f"Failed to decode message: {e}")
                except Exception as e:
                    logger.error(f"Error handling message: {e}")
                    
        except Exception as e:
            logger.error(f"Error in message listener: {e}")
        finally:
            self.running = False
    
    async def _handle_message(self, message: dict):
        """Handle incoming WebSocket messages."""
        event = message.get("event")
        payload = message.get("payload", {})
        
        if event == "killmail_update":
            system_id = payload.get("system_id")
            killmails = payload.get("killmails", [])
            timestamp = payload.get("timestamp")
            
            logger.info(f"🔥 New killmails in system {system_id}:")
            logger.info(f"   📊 Count: {len(killmails)}")
            logger.info(f"   ⏰ Timestamp: {timestamp}")
            
            for i, killmail in enumerate(killmails, 1):
                killmail_id = killmail.get("killmail_id")
                victim = killmail.get("victim", {})
                attackers = killmail.get("attackers", [])
                
                logger.info(f"   [{i}] Killmail ID: {killmail_id}")
                if victim:
                    victim_name = victim.get("character_name", "Unknown")
                    ship_name = victim.get("ship_type_name", "Unknown ship")
                    logger.info(f"       👤 Victim: {victim_name} ({ship_name})")
                
                if attackers:
                    logger.info(f"       ⚔️  Attackers: {len(attackers)}")
        
        elif event == "kill_count_update":
            system_id = payload.get("system_id")
            count = payload.get("count")
            logger.info(f"📊 Kill count update for system {system_id}: {count} kills")
        
        elif event == "phx_reply":
            # Handle command responses
            ref = message.get("ref")
            status = payload.get("status")
            response = payload.get("response", {})
            
            if status == "ok":
                logger.debug(f"✅ Command {ref} successful: {response}")
            else:
                logger.error(f"❌ Command {ref} failed: {response}")
    
    async def subscribe_to_systems(self, system_ids: List[int]) -> dict:
        """Subscribe to specific EVE Online systems."""
        if not self.websocket or not self.running:
            raise Exception("Not connected to WebSocket")
        
        message = {
            "topic": "killmails:lobby",
            "event": "subscribe_systems",
            "payload": {"systems": system_ids},
            "ref": 2
        }
        
        await self.websocket.send(json.dumps(message))
        self.subscriptions.update(system_ids)
        
        logger.info(f"✅ Subscribed to systems: {', '.join(map(str, system_ids))}")
        logger.info(f"📡 Total subscriptions: {len(self.subscriptions)}")
        
        return {"subscribed_systems": list(self.subscriptions)}
    
    async def unsubscribe_from_systems(self, system_ids: List[int]) -> dict:
        """Unsubscribe from specific EVE Online systems."""
        if not self.websocket or not self.running:
            raise Exception("Not connected to WebSocket")
        
        message = {
            "topic": "killmails:lobby",
            "event": "unsubscribe_systems",
            "payload": {"systems": system_ids},
            "ref": 3
        }
        
        await self.websocket.send(json.dumps(message))
        self.subscriptions.difference_update(system_ids)
        
        logger.info(f"❌ Unsubscribed from systems: {', '.join(map(str, system_ids))}")
        logger.info(f"📡 Remaining subscriptions: {len(self.subscriptions)}")
        
        return {"subscribed_systems": list(self.subscriptions)}
    
    async def get_status(self) -> dict:
        """Get current subscription status."""
        if not self.websocket or not self.running:
            raise Exception("Not connected to WebSocket")
        
        message = {
            "topic": "killmails:lobby",
            "event": "get_status",
            "payload": {},
            "ref": 4
        }
        
        await self.websocket.send(json.dumps(message))
        
        return {
            "subscription_id": self.subscription_id,
            "subscribed_systems": list(self.subscriptions),
            "connected": self.running
        }
    
    async def disconnect(self):
        """Disconnect from the WebSocket server."""
        self.running = False
        
        if self.websocket:
            # Send leave message
            leave_message = {
                "topic": "killmails:lobby",
                "event": "phx_leave",
                "payload": {},
                "ref": 5
            }
            
            try:
                await self.websocket.send(json.dumps(leave_message))
                await self.websocket.close()
            except:
                pass  # Ignore errors during cleanup
        
        logger.info("📴 Disconnected from WandererKills WebSocket")


async def example():
    """Example usage of the WandererKills WebSocket client."""
    client = WandererKillsClient('ws://localhost:4004')
    
    # Set up signal handler for graceful shutdown
    def signal_handler():
        logger.info("🛑 Shutdown signal received")
        asyncio.create_task(client.disconnect())
    
    # Handle SIGINT (Ctrl+C) and SIGTERM
    for sig in [signal.SIGINT, signal.SIGTERM]:
        signal.signal(sig, lambda s, f: signal_handler())
    
    try:
        # Connect to the server
        await client.connect()
        
        # Subscribe to some popular systems
        # Jita (30000142), Dodixie (30002659), Amarr (30002187)
        await client.subscribe_to_systems([30000142, 30002659, 30002187])
        
        # Get current status
        status = await client.get_status()
        logger.info(f"📋 Current status: {status}")
        
        # Keep running and listening for updates
        logger.info("🎧 Listening for killmail updates... Press Ctrl+C to stop")
        
        # Schedule unsubscription after 5 minutes
        async def delayed_unsubscribe():
            await asyncio.sleep(5 * 60)  # 5 minutes
            if client.running:
                await client.unsubscribe_from_systems([30000142])
        
        asyncio.create_task(delayed_unsubscribe())
        
        # Keep the client running until interrupted
        while client.running:
            await asyncio.sleep(1)
            
    except Exception as e:
        logger.error(f"❌ Client error: {e}")
    finally:
        await client.disconnect()


if __name__ == "__main__":
    # Run the example
    try:
        asyncio.run(example())
    except KeyboardInterrupt:
        logger.info("👋 Goodbye!")
    except Exception as e:
        logger.error(f"💥 Fatal error: {e}")
        sys.exit(1) 

File: examples/websocket_client.ex
defmodule WandererKills.WebSocketClient do
  @moduledoc """
  Elixir WebSocket client for WandererKills real-time killmail subscriptions.

  This example shows how to connect to the WandererKills WebSocket API
  from within Elixir to receive real-time killmail updates for specific
  EVE Online systems.

  ## Usage

      # Start the client
      {:ok, pid} = WandererKills.WebSocketClient.start_link([
        server_url: "ws://localhost:4004",
        systems: [30000142, 30002187]  # Jita, Amarr
      ])

      # Subscribe to additional systems
      WandererKills.WebSocketClient.subscribe_to_systems(pid, [30002659]) # Dodixie

      # Unsubscribe from systems
      WandererKills.WebSocketClient.unsubscribe_from_systems(pid, [30000142])

      # Get current status
      WandererKills.WebSocketClient.get_status(pid)

      # Stop the client
      WandererKills.WebSocketClient.stop(pid)
  """

  use GenServer
  require Logger

  alias Phoenix.Channels.GenSocketClient

  @behaviour GenSocketClient

  # Client API

  @doc """
  Start the WebSocket client.

  ## Options

    * `:server_url` - WebSocket server URL (required)
    * `:systems` - Initial systems to subscribe to (optional)
    * `:name` - Process name (optional)
  """
  @spec start_link(keyword()) :: GenServer.on_start()
  def start_link(opts) do
    {name, opts} = Keyword.pop(opts, :name, __MODULE__)
    GenServer.start_link(__MODULE__, opts, name: name)
  end

  @doc """
  Subscribe to additional EVE Online systems.
  """
  @spec subscribe_to_systems(pid() | atom(), [integer()]) :: :ok
  def subscribe_to_systems(client, system_ids) when is_list(system_ids) do
    GenServer.cast(client, {:subscribe_systems, system_ids})
  end

  @doc """
  Unsubscribe from EVE Online systems.
  """
  @spec unsubscribe_from_systems(pid() | atom(), [integer()]) :: :ok
  def unsubscribe_from_systems(client, system_ids) when is_list(system_ids) do
    GenServer.cast(client, {:unsubscribe_systems, system_ids})
  end

  @doc """
  Get current subscription status.
  """
  @spec get_status(pid() | atom()) :: {:ok, map()} | {:error, term()}
  def get_status(client) do
    GenServer.call(client, :get_status)
  end

  @doc """
  Stop the WebSocket client.
  """
  @spec stop(pid() | atom()) :: :ok
  def stop(client) do
    GenServer.stop(client)
  end

  # GenServer Callbacks

  @impl true
  def init(opts) do
    server_url = Keyword.fetch!(opts, :server_url)
    initial_systems = Keyword.get(opts, :systems, [])

    # Convert HTTP URL to WebSocket URL if needed
    websocket_url =
      server_url
      |> String.replace("http://", "ws://")
      |> String.replace("https://", "wss://")

    state = %{
      server_url: websocket_url,
      socket: nil,
      channel: nil,
      subscribed_systems: MapSet.new(),
      initial_systems: initial_systems,
      subscription_id: nil,
      connected: false
    }

    Logger.info("🚀 Starting WandererKills WebSocket client",
      server_url: websocket_url,
      initial_systems: initial_systems
    )

    # Connect asynchronously
    send(self(), :connect)

    {:ok, state}
  end

  @impl true
  def handle_info(:connect, state) do
    case connect_to_websocket(state) do
      {:ok, socket} ->
        Logger.info("✅ Connected to WandererKills WebSocket")

        # Join the killmails channel
        case join_channel(socket, state.initial_systems) do
          {:ok, channel} ->
            new_state = %{state |
              socket: socket,
              channel: channel,
              connected: true,
              subscribed_systems: MapSet.new(state.initial_systems)
            }

            Logger.info("📡 Joined killmails channel",
              initial_systems: state.initial_systems,
              systems_count: length(state.initial_systems)
            )

            {:noreply, new_state}

          {:error, reason} ->
            Logger.error("❌ Failed to join channel: #{inspect(reason)}")
            # Retry connection after delay
            Process.send_after(self(), :connect, 5_000)
            {:noreply, state}
        end

      {:error, reason} ->
        Logger.error("❌ Failed to connect to WebSocket: #{inspect(reason)}")
        # Retry connection after delay
        Process.send_after(self(), :connect, 5_000)
        {:noreply, state}
    end
  end

  @impl true
  def handle_cast({:subscribe_systems, system_ids}, %{connected: true} = state) do
    case push_to_channel(state.channel, "subscribe_systems", %{systems: system_ids}) do
      :ok ->
        new_subscriptions = MapSet.union(state.subscribed_systems, MapSet.new(system_ids))
        new_state = %{state | subscribed_systems: new_subscriptions}

        Logger.info("✅ Subscribed to additional systems",
          systems: system_ids,
          total_subscriptions: MapSet.size(new_subscriptions)
        )

        {:noreply, new_state}

      {:error, reason} ->
        Logger.error("❌ Failed to subscribe to systems: #{inspect(reason)}")
        {:noreply, state}
    end
  end

  def handle_cast({:subscribe_systems, _system_ids}, state) do
    Logger.warning("⚠️  Cannot subscribe: not connected to WebSocket")
    {:noreply, state}
  end

  @impl true
  def handle_cast({:unsubscribe_systems, system_ids}, %{connected: true} = state) do
    case push_to_channel(state.channel, "unsubscribe_systems", %{systems: system_ids}) do
      :ok ->
        new_subscriptions = MapSet.difference(state.subscribed_systems, MapSet.new(system_ids))
        new_state = %{state | subscribed_systems: new_subscriptions}

        Logger.info("❌ Unsubscribed from systems",
          systems: system_ids,
          remaining_subscriptions: MapSet.size(new_subscriptions)
        )

        {:noreply, new_state}

      {:error, reason} ->
        Logger.error("❌ Failed to unsubscribe from systems: #{inspect(reason)}")
        {:noreply, state}
    end
  end

  def handle_cast({:unsubscribe_systems, _system_ids}, state) do
    Logger.warning("⚠️  Cannot unsubscribe: not connected to WebSocket")
    {:noreply, state}
  end

  @impl true
  def handle_call(:get_status, _from, state) do
    status = %{
      connected: state.connected,
      server_url: state.server_url,
      subscription_id: state.subscription_id,
      subscribed_systems: MapSet.to_list(state.subscribed_systems),
      systems_count: MapSet.size(state.subscribed_systems)
    }

    Logger.info("📋 Current status", status)
    {:reply, {:ok, status}, state}
  end

  @impl true
  def handle_call(_msg, _from, state) do
    {:reply, {:error, :unknown_call}, state}
  end

  # Phoenix GenSocketClient Callbacks

  @impl GenSocketClient
  def handle_connected(transport, state) do
    Logger.debug("🔗 WebSocket transport connected", transport: inspect(transport))
    {:ok, state}
  end

  @impl GenSocketClient
  def handle_disconnected(reason, state) do
    Logger.warning("📡 WebSocket disconnected", reason: inspect(reason))

    # Update state and attempt reconnection
    new_state = %{state | connected: false, socket: nil, channel: nil}

    # Schedule reconnection
    Process.send_after(self(), :connect, 5_000)

    {:ok, new_state}
  end

  @impl GenSocketClient
  def handle_channel_closed(topic, payload, _transport, state) do
    Logger.warning("📺 Channel closed", topic: topic, payload: inspect(payload))

    # Update state
    new_state = %{state | channel: nil, connected: false}

    # Attempt to rejoin
    Process.send_after(self(), :connect, 2_000)

    {:ok, new_state}
  end

  @impl GenSocketClient
  def handle_message(topic, event, payload, _transport, state) do
    handle_channel_message(topic, event, payload, state)
  end

  @impl GenSocketClient
  def handle_reply(topic, ref, payload, _transport, state) do
    Logger.debug("📬 Channel reply",
      topic: topic,
      ref: ref,
      status: get_in(payload, ["status"])
    )

    # Handle join reply to extract subscription_id
    if topic == "killmails:lobby" and get_in(payload, ["status"]) == "ok" do
      subscription_id = get_in(payload, ["response", "subscription_id"])
      new_state = %{state | subscription_id: subscription_id}
      {:ok, new_state}
    else
      {:ok, state}
    end
  end

  # Private Helper Functions

  defp connect_to_websocket(state) do
    url = "#{state.server_url}/socket/websocket"

    socket_opts = [
      url: url,
      params: %{vsn: "2.0.0"}
    ]

    case GenSocketClient.start_link(__MODULE__, nil, socket_opts) do
      {:ok, socket} -> {:ok, socket}
      error -> error
    end
  end

  defp join_channel(socket, initial_systems) do
    join_params = if Enum.empty?(initial_systems) do
      %{}
    else
      %{systems: initial_systems}
    end

    case GenSocketClient.join(socket, "killmails:lobby", join_params) do
      {:ok, response} ->
        Logger.debug("📺 Joined channel", response: inspect(response))
        {:ok, socket}
      error -> error
    end
  end

  defp push_to_channel(socket, event, payload) when is_pid(socket) do
    case GenSocketClient.push(socket, "killmails:lobby", event, payload) do
      {:ok, _ref} -> :ok
      error -> error
    end
  end

  defp push_to_channel(_socket, _event, _payload) do
    {:error, :no_socket}
  end

  defp handle_channel_message("killmails:lobby", "killmail_update", payload, state) do
    system_id = payload["system_id"]
    killmails = payload["killmails"] || []
    timestamp = payload["timestamp"]
    is_preload = payload["preload"] || false

    if is_preload do
      Logger.info("📦 Preloaded killmails for system #{system_id}:",
        killmails_count: length(killmails),
        timestamp: timestamp,
        preload: true
      )
    else
      Logger.info("🔥 New real-time killmails in system #{system_id}:",
        killmails_count: length(killmails),
        timestamp: timestamp,
        preload: false
      )
    end

    # Process each killmail
    Enum.with_index(killmails, 1)
    |> Enum.each(fn {killmail, index} ->
      killmail_id = killmail["killmail_id"]
      victim = killmail["victim"] || %{}
      attackers = killmail["attackers"] || []

      character_name = victim["character_name"] || "Unknown"
      ship_name = victim["ship_type_name"] || "Unknown ship"
      kill_time = killmail["kill_time"] || killmail["killmail_time"] || "Unknown time"

      prefix = if is_preload, do: "📦", else: "🔥"

      Logger.info("   #{prefix} [#{index}] Killmail ID: #{killmail_id}",
        victim: character_name,
        ship: ship_name,
        kill_time: kill_time,
        attackers_count: length(attackers)
      )
    end)

    # You can add custom handling here:
    # - Store killmails in database
    # - Forward to other processes
    # - Trigger business logic
    # - Send notifications

    {:ok, state}
  end

  defp handle_channel_message("killmails:lobby", "kill_count_update", payload, state) do
    system_id = payload["system_id"]
    count = payload["count"]

    Logger.info("📊 Kill count update for system #{system_id}: #{count} kills")

    {:ok, state}
  end

  defp handle_channel_message(topic, event, payload, state) do
    Logger.debug("📨 Unhandled channel message",
      topic: topic,
      event: event,
      payload: inspect(payload)
    )

    {:ok, state}
  end
end

# Example usage module
defmodule WandererKills.WebSocketClient.Example do
  @moduledoc """
  Example usage of the WandererKills WebSocket client.

  Run this example with:

      iex> WandererKills.WebSocketClient.Example.run()
  """

  require Logger

  @doc """
  Run the WebSocket client example.
  """
  def run do
    Logger.info("🚀 Starting WandererKills WebSocket client example...")
    Logger.info("📋 This example demonstrates:")
    Logger.info("   1. Connecting with initial systems (Jita, Amarr)")
    Logger.info("   2. Receiving preloaded killmails for those systems")
    Logger.info("   3. Subscribing to additional systems (Dodixie)")
    Logger.info("   4. Receiving real-time killmail updates")

    # Start the client with some popular systems
    initial_systems = [30000142, 30002187]  # Jita, Amarr
    client_opts = [
      server_url: "ws://localhost:4004",
      systems: initial_systems,
      name: :wanderer_websocket_client
    ]

    Logger.info("🔌 Connecting to WebSocket with initial systems:",
      systems: initial_systems,
      system_names: ["Jita (30000142)", "Amarr (30002187)"]
    )

    case WandererKills.WebSocketClient.start_link(client_opts) do
      {:ok, pid} ->
        Logger.info("✅ WebSocket client started successfully")
        Logger.info("⏳ Watch for preloaded killmails to arrive shortly...")

        # Wait a bit for connection to establish and preload to complete
        Process.sleep(3_000)

        # Subscribe to additional systems after 5 seconds
        spawn(fn ->
          Process.sleep(5_000)
          Logger.info("📡 Adding Dodixie to subscriptions...")
          Logger.info("⏳ Watch for preloaded killmails from Dodixie...")
          WandererKills.WebSocketClient.subscribe_to_systems(pid, [30002659]) # Dodixie
        end)

        # Unsubscribe from Jita after 15 seconds
        spawn(fn ->
          Process.sleep(15_000)
          Logger.info("❌ Removing Jita from subscriptions...")
          WandererKills.WebSocketClient.unsubscribe_from_systems(pid, [30000142]) # Jita
        end)

        # Get status after 20 seconds
        spawn(fn ->
          Process.sleep(20_000)
          case WandererKills.WebSocketClient.get_status(pid) do
            {:ok, status} ->
              Logger.info("📋 Current client status:",
                connected: status.connected,
                systems_count: status.systems_count,
                subscribed_systems: status.subscribed_systems
              )
            {:error, reason} ->
              Logger.error("❌ Failed to get status: #{inspect(reason)}")
          end
        end)

        # Keep the example running
        Logger.info("")
        Logger.info("🎧 Client is now listening for killmail updates...")
        Logger.info("💡 You can interact with it using:")
        Logger.info("   WandererKills.WebSocketClient.subscribe_to_systems(:wanderer_websocket_client, [system_ids])")
        Logger.info("   WandererKills.WebSocketClient.unsubscribe_from_systems(:wanderer_websocket_client, [system_ids])")
        Logger.info("   WandererKills.WebSocketClient.get_status(:wanderer_websocket_client)")
        Logger.info("   WandererKills.WebSocketClient.stop(:wanderer_websocket_client)")
        Logger.info("")
        Logger.info("📦 Preloaded killmails have a 📦 icon")
        Logger.info("🔥 Real-time killmails have a 🔥 icon")

        {:ok, pid}

      {:error, reason} ->
        Logger.error("❌ Failed to start WebSocket client: #{inspect(reason)}")
        {:error, reason}
    end
  end

  @doc """
  Stop the example client.
  """
  def stop do
    case Process.whereis(:wanderer_websocket_client) do
      nil ->
        Logger.info("ℹ️  WebSocket client is not running")
        :ok

      pid ->
        WandererKills.WebSocketClient.stop(pid)
        Logger.info("🛑 WebSocket client stopped")
        :ok
    end
  end

  @doc """
  Simple example that connects with initial systems and shows preload behavior.

  This demonstrates the most common use case: connecting to a few systems
  and immediately receiving cached killmail data.
  """
  def simple_example do
    Logger.info("🚀 Simple WebSocket client example")
    Logger.info("📋 Connecting to Jita and Amarr with preload...")

    # Connect with initial systems
    {:ok, pid} = WandererKills.WebSocketClient.start_link([
      server_url: "ws://localhost:4004",
      systems: [30000142, 30002187],  # Jita, Amarr
      name: :simple_client
    ])

    Logger.info("✅ Connected! Watch for preloaded killmails...")

    # Wait and show status
    Process.sleep(5_000)

    case WandererKills.WebSocketClient.get_status(pid) do
      {:ok, status} ->
        Logger.info("📋 Client status: #{status.systems_count} systems, connected: #{status.connected}")
      {:error, _} ->
        Logger.warning("⚠️  Could not get client status")
    end

    Logger.info("🎧 Listening for real-time updates... (Ctrl+C to stop)")

    # Keep running until stopped
    Process.sleep(:infinity)
  end
end
